docs/book/src/tasks/experimental-features/cluster-class/write-clusterclass.md

```diff
@@ -676,6 +677,102 @@ individually.
 
 </aside>
 
+## JSON patches tips & tricks
+
+JSON patches specification [RFC6902] requires that the target of
+add operation must exist.
+
+As a consequence ClusterClass authors should pay special attention when the following
+conditions apply in order to prevent errors when a patch is applied:
+
+* the patch is a JSON patch
+* the patch tries to `add` a value to a **slice**
+* the slice was defined with `omitempty`
+* the slice currently does not exist
+
+A workaround in this particular case is to create the slice in the patch instead of adding it to the non-existing one.
+When creating the slice, existing values would be overwritten so this should only be used when it does not exist.
+
+The following example shows both cases to consider while writing a patch for adding a value to a slice.
+This patch targets to add a file to the `files` slice of a `KubeadmConfigTemplate` which has [omitempty](https://github.com/kubernetes-sigs/cluster-api/blob/main/bootstrap/kubeadm/api/v1beta1/kubeadmconfig_types.go#L54) set.
+
+{{#tabs name:"tab-configuration-patches" tabs:"Add to existing slice,Create slice"}}
+{{#tab Add to existing slice}}
+
+This patch **requires** the key `.spec.template.spec.files` to exist to succeed.
+
+```yaml
+apiVersion: cluster.x-k8s.io/v1beta1
+kind: ClusterClass
+metadata:
+  name: my-clusterclass
+spec:
+  ...
+  patches:
+  - name: add file
+    definitions:
+    - selector:
+        apiVersion: bootstrap.cluster.x-k8s.io/v1beta1
+        kind: KubeadmConfigTemplate
+      jsonPatches:
+      - op: add
+        path: /spec/template/spec/files/-
+        value:
+          content: Some content.
+          path: /some/file
+---
+apiVersion: bootstrap.cluster.x-k8s.io/v1beta1
+kind: KubeadmConfigTemplate
+metadata:
+  name: "quick-start-default-worker-bootstraptemplate"
+spec:
+  template:
+    spec:
+      ...
+      files:
+      - content: Some other content
+        path: /some/other/file
+```
+
+{{#/tab }}
+{{#tab Create slice}}
+
+This patch would **overwrite** an existing slice at `.spec.template.spec.files`.
```

Comment:
I think it would fail because the JSON patch lib would detect that it can't add an already existing array (only with op: replace it would overwrite it)

Reply:
The library used in clusterclass for JSON patching would overwrite it.

```
apiVersion: bootstrap.cluster.x-k8s.io/v1beta1
kind: KubeadmConfigTemplate
spec:
  templates:
    spec:
      files:
      - content: Old
        path: /some
```

And the patch below it would result in
```
apiVersion: bootstrap.cluster.x-k8s.io/v1beta1
kind: KubeadmConfigTemplate
spec:
  templates:
    spec:
      files:
      - content: Some content.
        path: /some/file
```

I will verify that again tomorrow.

I agree with you, normally it should be invalid, e.g. https://json-schema-validator.herokuapp.com/jsonpatch.jsp would throw an error.

Reply:
Wow, I thought I remembered having that error. But it's absolutely possible that I'm wrong.

Just one fyi in case that's somehow relevant. I think during some version bump the jsonpatch lib (https://github.com/evanphx/json-patch) got more strict. I think it was from 4 to 5.

Reply:
I'm pretty sure I had been on main at the time of testing. But I will verify that.

Reply:
Ah, its the library which actually allows it https://go.dev/play/p/bT8wcbh7BSQ

Reply:
Okay then you're right :). Maybe it was the old version which was more strict